include/mega/types.h

```diff
@@ -440,14 +441,67 @@ typedef enum { TRANSFERSTATE_NONE = 0, TRANSFERSTATE_QUEUED, TRANSFERSTATE_ACTIV
                TRANSFERSTATE_RETRYING, TRANSFERSTATE_COMPLETING, TRANSFERSTATE_COMPLETED,
                TRANSFERSTATE_CANCELLED, TRANSFERSTATE_FAILED } transferstate_t;
 
-struct Notification
+template<class T>
+struct ThreadSafeDeque
 {
-    dstime timestamp;
-    string path;
-    LocalNode* localnode;
-};
+    // Just like a deque, but thread safe so that a separate thread can receive filesystem notifications as soon as they are available.
+    // When we try to do that on the same thread, the processing of queued notifications is too slow so more notifications bulid up than
+    // have been processed, so each time we get the outstanding ones from the buffer we gave to the OS, we need to give it an even 
+    // larger buffer to write into, otherwise it runs out of space before this thread is idle and can get the next batch from the buffer.
+protected:
+    std::deque<T> mNotifications;
+    std::mutex m;
+
+public:
+
+    bool peekFront(T& t) 
+    { 
+        std::lock_guard g(m);
+        if (!mNotifications.empty())
+        {
+            t = mNotifications.front();
+            return true; 
+        }
+        return false;
+    }
+
+    bool popFront(T& t) 
+    { 
+        std::lock_guard g(m);
+        if (!mNotifications.empty())
+        {
+            t = std::move(mNotifications.front());
+            mNotifications.pop_front(); 
+            return true; 
+        }
+        return false;
+    }
+
+    void unpopFront(const T& t) 
```

Comment:
Why unpopFront instead of pushFront?

Reply:
For this usage, the item put on the front should only be the item that was just taken off, and also calling it pushFront would make it seem equal to pushBack meaning you could use it in either direction, but pushing back and popping front is the envisaged use here, I thought this naming would make it easier to see which was the intended direction.  thanks!